/*
 * The communication protocol between S2E and the symbex plugin.
 */

package chef.data;

// A symbolic name -> value assignment.
message VariableAssignment {
  required string name = 1;
  required bytes value = 2;
}

message InputAssignment {
  repeated VariableAssignment var_assignment = 1;
}

message ForkPoint {
  optional uint32 depth = 1;
  optional fixed64 target_pc = 2;
  optional uint32 branch_index = 3;
}


message ExecutionDistance {
	required uint32 path_distance = 1;
	required uint32 cfg_distance = 2;
}


// A test case comprising multiple symbolic assignments
message TestCase {
  required uint64 time_stamp = 10;

  // Input attributes

  required InputAssignment input = 1;
    // The seed test cases collected with this test case
  repeated InputAssignment seed_input = 9;

  // Execution attributes

  optional uint64 high_level_path_id = 8;
  repeated ForkPoint fork_point = 7;

  // FIXME: Shady metrics, should remove in the future
  optional uint32 dist_to_uncovered = 14;
  optional ExecutionDistance tree_divergence = 12;
  optional ExecutionDistance cfg_divergence = 13;

  // Output attributes

  optional bytes output = 11;
}
